internal/search/crawler/github/crawler.go

```diff
@@ -0,0 +1,420 @@
+// Package github implements the crawler.Crawler interface, getting data its
+// from the Github search API.
+package github
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math"
+	"net/http"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"sigs.k8s.io/kustomize/internal/search/doc"
+)
+
+var logger = log.New(os.Stdout, "Github Crawler: ",
+	log.LstdFlags|log.LUTC|log.Llongfile)
+
+// Implements crawler.Crawler.
+type githubCrawler struct {
+	rc    RequestConfig
+	query Query
+}
+
+func NewCrawler(
+	accessToken string, retryCount uint64, query Query) githubCrawler {
+
+	return githubCrawler{
+		rc: RequestConfig{
+			perPage:     githubMaxPageSize,
+			retryCount:  retryCount,
+			accessToken: accessToken,
+		},
+		query: query,
+	}
+}
+
+// Implements crawler.Crawler.
+func (gc githubCrawler) Crawl(
+	ctx context.Context, output chan<- *doc.KustomizationDocument) error {
+
+	// Range finding will be added in the next PR to make this one
+	// simpler/shorter. It would return multiple search queries for the
+	// Github API such that all documents can be retrieved. This is
+	// required since Github returns a max of 1000 results per query, so
+	// multiple queries that split the search space into chunks of 1000
+	// kustomization files is required.
+	ranges := []string{gc.query.String()}
+
+	// Query each range for files.
+	errs := make(multiError, 0)
+	for _, query := range ranges {
+		err := ProcessQuery(ctx, gc.rc, query, output)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return errs
+}
+
+// ProcessQuery follows all of the pages in a query, and updates/adds the
+// documents from the crawl to the datastore/index.
+func ProcessQuery(ctx context.Context, rc RequestConfig, query string,
```

Comment:
Change `ProcessQuery` to a priviate function `processQuery` if it's only used in this package.